chain/chain/src/chain.rs

```diff
@@ -2891,98 +2900,187 @@ impl<'a> ChainUpdate<'a> {
-                            false,
-                            self.states_to_patch.take(),
-                        )
-                        .map_err(|e| ErrorKind::Other(e.to_string()))?;
-
-                    self.chain_store_update.save_trie_changes(apply_result.trie_changes);
-                    *new_extra.state_root_mut() = apply_result.new_root;
-
-                    self.chain_store_update.save_chunk_extra(&block.hash(), shard_id, new_extra);
+                    let runtime_adapter = self.runtime_adapter.clone();
+                    let block_hash = block.hash().clone();
+                    let challenges_result = block.header().challenges_result().clone();
+                    let block_timestamp = block.header().raw_timestamp();
+                    let gas_price = block.header().gas_price();
+                    let random_seed = *block.header().random_value();
+                    let height = block.header().height();
+                    let prev_block_hash = prev_block.hash().clone();
+                    #[cfg(feature = "sandbox")]
+                    let states_to_patch = self.states_to_patch.take();
+                    #[cfg(not(feature = "sandbox"))]
+                    let _ = self.states_to_patch;
+
+                    dif_height_handlers.push(thread::spawn(
+                        move || -> Result<(ShardId, ApplyTransactionResult), Error> {
+                            match runtime_adapter.apply_transactions(
+                                shard_id,
+                                new_extra.state_root(),
+                                height,
+                                block_timestamp,
+                                &prev_block_hash,
+                                &block_hash,
+                                &[],
+                                &[],
+                                new_extra.validator_proposals(),
+                                gas_price,
+                                new_extra.gas_limit(),
+                                &challenges_result,
+                                random_seed,
+                                false,
+                                false,
+                                #[cfg(feature = "sandbox")]
+                                states_to_patch,
+                                #[cfg(not(feature = "sandbox"))]
+                                None,
+                            ) {
+                                Ok(result) => Ok((shard_id, result)),
+                                Err(err) => Err(err),
+                            }
+                        },
+                    ));
                 }
             }
         }
 
-        Ok(())
+        let block_hash = block.hash().clone();
+        let prev_block_hash = prev_block.hash().clone();
+        Ok(move |chain_update: &mut ChainUpdate| -> Result<(), Error> {
+            for result in same_height_handlers {
+                let res = result.join();
```

Comment:
I'd just

```suggestion
                let res = result.join().unwrap();
```

here. 

The panic message is displayed in the thread that panicked, so here we can just propagate without additional logging or handling. I'd even say that it's a minor problem in the API that, by default join returns a Result, it should have paniked by default: https://matklad.github.io/2020/12/12/notes-on-lock-poisoning.html.